README.markdown
# TimePicker: A simple solution for picking time


## About


**jquery-timepicker** a very simple jquery plugin, originally developed by [Jason Huck](http://www.corefive.com/), and improved by [Anuj Dutta](http://www.andhapp.com/blog) by adding unit tests and improving the overall readability and standard of the code. Code was originally found at [Google's source code repository](http://code.google.com/p/jquery-timepicker). 


## Inspiration

Simple. Scratch my itch. I could not find a good solution then found Jason's plugin (also listed on the jquery plugin site). It was easy to use and I decided to improve the code and add tests to it.


## Usage

Again, simple. It hides an input box and displays three (or two based on supplied options) select boxes, one each for hour, mins and time of day (am or pm). All, you need to do is attach the timepicker function to the input box that will be used for time input.

Example: Your HTML code looks like this:
    
    <div id="timeContainer">
        <input type="text" id="select_time" value="">
    </div>
  
and your javascript code should look like this:
  
    $('input#select_time').timepicker();
  
Please make sure the input box is inside a div since the select boxes are appended to the parent div.

Works with **jquery 1.7.2**.

### Options

1. You can supply the option to display a 24-hr clock and not display time-of-day selection box by using the following javascript:
  
    $('input#select_time').timepicker({clock12:false});

2. You can make the timepicker appear in a popup box by passing in
   inPopup:true. Default is false.
  
    $('input#select_time').timepicker({inPopup:true});


If you get stuck, just look at the tests. There is also an example
index.html file in spec/integration.

## Specs

The plugin is tested using [Evergreen](http://github.com/jnicklas/evergreen), a behaviour driven testing framework for javascript built on top of Jasmine. To run the specs, clone the repository, install required gems and then run the following:

    bundle exec evergreen run    

## Versioning

The code follows seamntic versioning and is at 1.1.0 at the moment.


## Roadmap

* Bug in Safari in parseInt. You have to specify the radix. This might
  be fixed now, as I didn't see it again while testing.
* Add timezones
* Create a gem that can be packged just like jquery-rails (although, it might create bugs since this plugin depends on one version and jquery-rails might depend on completely different)
* Add locale String, only for am/pm strings.



Changelog.md
# Changelng

## 29 April 2012

* Ensured that the timepicker has the correct position when it fades
  in. [Bug fix]

## 17 April 2012

* Added VERSION

## 16 April 2012

* Added code to allow the timepicker to be added into a popup container

## 05 April 2012

* Updated README

## 04 April 2012

* Moved all the specs to work with Evergreen.


test/unit/timepicker.js
// <![CDATA[
		
	describe('On binding input type "text" to timepicker', {
		
		before_each: function() {
			$('input#select_time').timepicker();
		},
		
		'Should change the input box to type hidden': function() {
			value = $('input#select_time').attr('type');
			
			expect(value).should_be("hidden");
		},
		
		'Should insert select box for hour': function() {
			value = $('select#h_select_time').is(":visible");
			num_options_in_hr_select = $('#h_select_time').children().length;
			
			expect(value).should_be(true);
			expect(num_options_in_hr_select).should_be(12);
		},
		
		'Should insert select box for minutes': function() {
			value = $('#m_select_time').is(":visible");
			
			expect(value).should_be(true);			
		},
		
		'Should insert select box for am & pm by default': function() {
			value = $('#p_select_time').is(":visible");
			
			expect(value).should_be(true);			
		},
		
		after_each: function() {
			$("select[id$='_select_time']").each(function() {
				$(this).remove();
			});
		}				
	})
	
	describe('On passing options to timepickr', {
		
		'Should not insert select box for am & pm if 12-hr clock == false': function() {
			$('input#select_time').timepicker({clock12:false});
			num_options_in_hr_select = $('#h_select_time').children().length;
			value = $('#p_select_time').is(":visible");
			
			expect(value).should_be(false);
			expect(num_options_in_hr_select).should_be(24);
		},
		
		after_each: function() {
			$("select[id$='_select_time']").each(function() {
				$(this).remove();
			});
		}				
		
	})
	
	describe('On selecting values from the select boxes with am/pm', {
		
		// NOTE: Using after_each here works in a weird fashion. Therefore, calling it explicitly after this example.
		'Should retreive values from select boxes and insert into the hidden input box': function() {
			$('input#select_time').timepicker();
			$('#h_select_time').attr("value", "09");
			$('#m_select_time').attr("value", "15");
			$('#ap_select_time').attr("value", "am");
			
			$('#h_select_time').trigger("change");
			$('#m_select_time').trigger("change");
			$('#ap_select_time').trigger("change");
			value = $('input#select_time').attr("value");
			
			expect(value).should_be("09:15am");						
		},
		
		after_each: function() {
			$("select[id$='_select_time']").each(function() {
				$(this).remove();
			});
		}				
		
	})	
	
	describe('On selecting values from the select boxes without am/pm', {
		
		// NOTE: Using after_each here works in a weird fashion. Therefore, calling it explicitly after this example.
		'Should retreive values from select boxes and insert into the hidden input box without am/pm': function() {
			$('input#select_time').timepicker({clock12:false});
			$('#h_select_time').attr("value", "21");
			$('#m_select_time').attr("value", "15");
			
			$('#h_select_time').trigger("change");
			$('#m_select_time').trigger("change");
			value = $('input#select_time').attr("value");
			
			expect(value).should_be("21:15");
		}, 
		
		after_each: function() {
			$("select[id$='_select_time']").each(function() {
				$(this).remove();
			});
		}				
		
	})	
	
					
// ]]>

spec/javascripts/timepicker_spec.js
require('/javascripts/jquery.js')
require('/javascripts/jquery.timepicker.js')

function createAndAppendElement(id, parent) {
  selectTimeInput = $('<input>', {id: id})
  parent.append(selectTimeInput)
  return selectTimeInput;
}

function hour_select_element() {
  return $('#h_' + id)
}

function minute_select_element() {
  return $('#m_' + id)
}

function am_pm_select_element() {
  return $('#p_' + id)
}

describe('Binding input type text with timepicker', function() {
  beforeEach(function() {
    test = $('#test')
    id = 'selectTime'
  })
  describe('with default options', function() {
    beforeEach(function() {
      selectTimeInput = createAndAppendElement(id, test)

      selectTimeInput.timepicker()
      parentContainer = selectTimeInput.parent(0)

      hour_select = hour_select_element()
      minute_select = minute_select_element()
      am_pm_select = am_pm_select_element()
    })
    it('should make input box hidden', function() {
      expect(selectTimeInput.attr('type')).toEqual('hidden')
    })
    it('should insert select box for hour', function() {
      num_children = hour_select.children('option').length

      expect(hour_select.is(":visible")).toEqual(true)
      expect(num_children).toEqual(12)
    })
    it('should insert select box for minutes', function() {
      num_children = minute_select.children('option').length

      expect(minute_select.is(":visible")).toEqual(true)
      expect(num_children).toEqual(4) //00, 15, 30, 45
    })
    it('should insert select box for am/pm', function() {
      num_children = am_pm_select.children('option').length

      expect(am_pm_select.is(":visible")).toEqual(true)
      expect(num_children).toEqual(2)
    })
    it('should insert selected values into hidden input box', function() {
      hour_select.attr("value", "09").trigger("change");
      minute_select.attr("value", "15").trigger("change");
      am_pm_select.attr("value", "am").trigger("change")
			
      expect(selectTimeInput.attr("value")).toEqual('09:15am')
    })
    it('should insert the hour, minutes and am/pm in input box parent container', function() {
      expect(hour_select.parent(0).attr('id')).toEqual('test')
      expect(minute_select.parent(0).attr('id')).toEqual('test')
      expect(am_pm_select.parent(0).attr('id')).toEqual('test')
    })
  })
  describe('with user options', function() {
    describe('inPopup:true', function() {
      beforeEach(function() {
        id = id + 'UserOptionsPopup'

        selectTimeInputUserOptionsPopup = createAndAppendElement(id, test)

        selectTimeInputUserOptionsPopup.timepicker({inPopup:true})

        hour_select = hour_select_element()
        minute_select = minute_select_element()
        am_pm_select = am_pm_select_element()

        parentContainer = hour_select.parent(0)
      })
      it('should be hidden by default', function() {
        expect(parentContainer.is(":visible")).toEqual(false)
      })
      it('should be positioned absolute by default', function() {
        expect(parentContainer.css('position')).toEqual('absolute')
      })
      it('should be positioned just below the input box', function() {
        expected_top = selectTimeInputUserOptionsPopup.offset().top +
                       selectTimeInputUserOptionsPopup.outerHeight() + 'px'

        expected_left = selectTimeInputUserOptionsPopup.offset().left + 'px'

        expect(parentContainer.css('top')).toEqual(expected_top)
        expect(parentContainer.css('left')).toEqual(expected_left)
      })
      it('should have class=timepicker and id=timepicker', function() {
        expect(parentContainer.attr('id')).toEqual('timepicker')
        expect(parentContainer.attr('class')).toEqual('timepicker')
      })
      it('should insert the hour, minutes and am/pm in timepicker container', function() {
        expect(hour_select.parent(0).attr('id')).toEqual('timepicker')
        expect(minute_select.parent(0).attr('id')).toEqual('timepicker')
        expect(am_pm_select.parent(0).attr('id')).toEqual('timepicker')
      })
      it('should show the timepicker container on input box focus', function() {
        selectTimeInputUserOptionsPopup.trigger("focus")

        expect(parentContainer.is(":visible")).toEqual(true)
      })
      it('should hide the timepicker container when user clicks outside of input box or the timepicker container', function() {
        runs(function() {
          selectTimeInputUserOptionsPopup.trigger("focus")
        })

        waits(500)

        runs(function(){
          test.trigger("click")
          expect(parentContainer.is(":visible")).toEqual(false)
        })
      })
      it('should not hide the timepicker container when user clicks in timepicker container', function() {
        runs(function() {
          selectTimeInputUserOptionsPopup.trigger("focus")
        })

        waits(500)

        runs(function(){
          parentContainer.trigger("click")
          expect(parentContainer.is(":visible")).toEqual(true)
        })
      })
      it('should not hide the timepicker container when user clicks in input box', function() {
        runs(function() {
          selectTimeInputUserOptionsPopup.trigger("focus")
        })

        waits(500)

        runs(function(){
          selectTimeInputUserOptionsPopup.trigger("focus")
          expect(parentContainer.is(":visible")).toEqual(true)
        })
      })
      it('should not hide the timepicker container when user selects time values', function() {
        runs(function() {
          selectTimeInputUserOptionsPopup.trigger("focus")
        })

        waits(500)

        runs(function(){
          hour_select_element().trigger("change")
        })

        waits(100)

        runs(function(){
          expect(parentContainer.is(":visible")).toEqual(true)
        })
      })
      afterEach(function() {
      })
    })
    describe('clock12:false', function() {
      beforeEach(function() {
        id = id + 'UserOptionsClock'
        selectTimeInputUserOptionsClock = createAndAppendElement(id, test)

        selectTimeInputUserOptionsClock.timepicker({clock12:false})
        hour_select = hour_select_element()
        minute_select = minute_select_element()
        am_pm_select = am_pm_select_element()
      })
      it('should not insert select box for am/pm', function() {
        expect(am_pm_select.is(":visible")).toEqual(false)
      })
      it('should insert hour select box with 24 values for a 24-hour clock', function() {
        num_children = hour_select.children('option').length

        expect(hour_select.is(":visible")).toEqual(true)
        expect(num_children).toEqual(24)
      })
      it('should insert values selected into the input box', function() {
        hour_select.attr("value", "21").trigger("change");
        minute_select.attr("value", "15").trigger("change");

        expect(selectTimeInputUserOptionsClock.attr("value")).toEqual('21:15')
      })
    })
  })
})


public/javascripts/jquery.timepicker.js
/* jQuery timepicker
 * Replaces a single text input with a set of select boxes for hour, minute, and am/pm (optional)
 *
 * Original copyright (c) 2007 Jason Huck/Core Five Creative (http://www.corefive.com/)
 * Dual licensed under the MIT (http://www.opensource.org/licenses/mit-license.php) 
 * and GPL (http://www.opensource.org/licenses/gpl-license.php) licenses.
 *
 * Improved by Anuj Dutta (@andhapp) by adding specs and refactoring the code to remove duplication 
 * and improve code standards
 *
 * Version 1.1.1
 *
 */

;(function($){	
	$.fn.timepicker = function(user_options) {
		
    var elem = this
    var CSS_CLASS = 'timepicker'
    var PARENT_CONTAINER_ID = 'timepicker'
    var ELEMENT_CSS_CLASS = 'timepickerElement'

		return this.each(function() {
			var options = $.extend({}, $.timepicker.selectableValues, $.timepicker, user_options)

      var timePicker = new TimePicker(elem, options)
      var htmlResult = timePicker.process()

      if(options.inPopup) {
        var popUpContainer = createPopupContainer()
        popUpContainer.append(htmlResult)

        elem.focusin(function() { focusInHandler(elem, popUpContainer) })
      } else {
        makeHidden(this) // Explicity setting the type property
        elem.parent().append(htmlResult)
      }

      timePicker.handleEvents()
			
		})

    function focusInHandler(elem, object) {
      setPosition(elem, object)

      object.fadeIn(400, function() {
        $(document).bind('click', function(event) {
          if(canHidePopupContainer(event)) {
            object.hide()
            $(document).unbind('click')
          }
        })
      })

    }

    function canHidePopupContainer(event) {
      function isPopupContainer(event) {
        return $(event.target).attr('class') === CSS_CLASS
      }

      function isInputBox(event) {
        return $(event.target)[0] === elem[0]
      }

      function isTimepickerDropdowns(event) {
        return $(event.target).attr('class') && $(event.target).attr('class').indexOf(ELEMENT_CSS_CLASS) > 0
      }

      return (
        ! ( isPopupContainer(event) ||
            isInputBox(event) ||
            isTimepickerDropdowns(event))
      )
    }

    function createPopupContainer() {

      var popUpContainer = {}

      if ($("#" + PARENT_CONTAINER_ID).length > 0) {
        popUpContainer = $('#' + PARENT_CONTAINER_ID)
      } else {
        popUpContainer = $('<div>', {id: PARENT_CONTAINER_ID, style:'display:none; position:absolute;', 'class': CSS_CLASS})
      }

      $('body').append(popUpContainer)

      return setPosition(elem, popUpContainer)
    }

    function setPosition(source, object) {
      var top = source.offset().top + source.outerHeight()
      var left = source.offset().left
      object.css({top: top, left: left})
      return object
    }

    function makeHidden(object) {
      object.type = "hidden"
    }
		
	}

  function TimePicker(elem, options) {
    this.element = elem
    this.options = options

    this.handleEvents = function() {
      var self = this
      $('.timepickerElement').change(function() {
        var final_value = self.hours.selectedValue() + ':' + self.minutes.selectedValue()

        if(self.ampm) {
          final_value += self.ampm.selectedValue()
        }

        self.element.val(final_value)
      })
    }

    this.process = function() {
      var result = ''
      var id = this.element.attr('id')

      if(this.options.clock12) {
        this.hours = new Hours(id, this.options.hrs12)
        this.ampm = new AmPm(id, this.options.ap)
      } else {
        this.hours = new Hours(id, this.options.hrs24)
      }

      this.minutes = new Minutes(id, this.options.mins)

      result += this.hours.createDOMElementHTML()
      result += this.minutes.createDOMElementHTML()

      if(this.ampm) {
        result += this.ampm.createDOMElementHTML()
      }

      return result
    }

    function createSelectElementHTML(id, className, source) {
      var source = source.split(',')
      var result = '<select id="' + className + '_' + id + '" class="' + className + ' timepickerElement">'

      $.each(source, function(index, value) {
        result += '<option value="' + value + '"'
        result += '>' + value + '</option>'
      })
      result += '</select>'

      return result;
    }

    function Hours(id, value) {
      this.id = id
      this.value = value
      this.className = 'h'

      this.createDOMElementHTML = function() {
        return createSelectElementHTML(this.id, this.className, this.value)
      }
      this.domElement = function() {
        return $('#' + this.className + '_' + this.id)
      }
      this.selectedValue = function() {
        return this.domElement().attr('value')
      }
    }

    function Minutes(id, value) {
      this.id = id
      this.value = value
      this.className = 'm'

      this.createDOMElementHTML = function() {
        return createSelectElementHTML(this.id, this.className, this.value)

      }
      this.domElement = function() {
        return $('#' + this.className + '_' + this.id)
      }
      this.selectedValue = function() {
        return this.domElement().attr('value')
      }
    }

    function AmPm(id, value) {
      this.id = id
      this.value = value
      this.className = 'p'

      this.createDOMElementHTML = function() {
        return createSelectElementHTML(this.id, this.className, this.value)
      }
      this.domElement = function() {
        return $('#' + this.className + '_' + this.id)
      }
      this.selectedValue = function() {
        return this.domElement().attr('value')
      }
    }

  }

  $.timepicker = {
		selectableValues: {
			ap: 		"am,pm",
			mins: 	"00,15,30,45",
			hrs12: 	"01,02,03,04,05,06,07,08,09,10,11,12",
			hrs24: 	"00,01,02,03,04,05,06,07,08,09,10,11,12,13,14,15,16,17,18,19,20,21,22,23",
		},
		clock12: true,
		inPopup: false
	}

})(jQuery)
